buffer/src/main/java/io/netty/buffer/PoolArena.java

```diff
@@ -148,114 +129,83 @@ protected PoolArena(PooledByteBufAllocator parent, int pageSize,
-        int tableIdx = 0;
-        int i = normCapacity >>> 10;
-        while (i != 0) {
-            i >>>= 1;
-            tableIdx ++;
+        if (sizeIdx <= smallMaxSizeIdx) {
+            tcacheAllocateSmall(cache, buf, reqCapacity, sizeIdx);
+        } else if (sizeIdx < nSizes) {
+            tcacheAllocateNormal(cache, buf, reqCapacity, sizeIdx);
+        } else {
+            int normCapacity = directMemoryCacheAlignment > 0
+                    ? normalizeSize(reqCapacity) : reqCapacity;
+            // Huge allocations are never served via the cache so just call allocateHuge
+            allocateHuge(buf, normCapacity);
         }
-        return tableIdx;
     }
 
-    // capacity < pageSize
-    boolean isTinyOrSmall(int normCapacity) {
-        return (normCapacity & subpageOverflowMask) == 0;
-    }
+    private void tcacheAllocateSmall(PoolThreadCache cache, PooledByteBuf<T> buf, final int reqCapacity,
+                                     final int sizeIdx) {
 
-    // normCapacity < 512
-    static boolean isTiny(int normCapacity) {
-        return (normCapacity & 0xFFFFFE00) == 0;
-    }
+        if (cache.allocateSmall(this, buf, reqCapacity, sizeIdx)) {
+            // was able to allocate out of the cache so move on
+            return;
+        }
 
-    private void allocate(PoolThreadCache cache, PooledByteBuf<T> buf, final int reqCapacity) {
-        final int normCapacity = normalizeCapacity(reqCapacity);
-        if (isTinyOrSmall(normCapacity)) { // capacity < pageSize
-            int tableIdx;
-            PoolSubpage<T>[] table;
-            boolean tiny = isTiny(normCapacity);
-            if (tiny) { // < 512
-                if (cache.allocateTiny(this, buf, reqCapacity, normCapacity)) {
-                    // was able to allocate out of the cache so move on
-                    return;
-                }
-                tableIdx = tinyIdx(normCapacity);
-                table = tinySubpagePools;
-            } else {
-                if (cache.allocateSmall(this, buf, reqCapacity, normCapacity)) {
-                    // was able to allocate out of the cache so move on
-                    return;
-                }
-                tableIdx = smallIdx(normCapacity);
-                table = smallSubpagePools;
+        /**
+         * Synchronize on the head. This is needed as {@link PoolChunk#allocateSubpage(int)} and
+         * {@link PoolChunk#free(long)} may modify the doubly linked list as well.
+         */
+        final PoolSubpage<T> head = smallSubpagePools[sizeIdx];
+        synchronized (head) {
+            final PoolSubpage<T> s = head.next;
+            if (s != head) {
+                assert s.doNotDestroy && s.elemSize == sizeIdx2size(sizeIdx);
+                long handle = s.allocate();
+                assert handle >= 0;
+                s.chunk.initBufWithSubpage(buf, null, handle, reqCapacity, cache);
+                incSmallAllocation();
```

Comment:
I think it would be better if we could move this out of the `synchronized(...)` block

Reply:
@yaroot did you see this ?

Reply:
@normanmaurer Do you mean move incSmallAllocation() out of the {}block？

Reply:
`s.allocate()`may change the structure of linked list, we need to synchronized on the head.

Reply:
@yuanrw I was only talking about `incSmallAllocation()`